### ISSUE TITLE
When stopping the raft service, it may cause a crash.

### ISSUE THREAD

**Issue by monadbobo** [vesoft-inc/nebula#1077]
When stopping the raft service, it may cause a crash.

Below is the output of AddressSanitizer:

```
READ of size 8 at 0x616000330080 thread T145 (ThriftIO0)
    #0 0xb1230b in auto nebula::thrift::ThriftClientManager<nebula::raftex::cpp2::RaftexServiceAsyncClient>::client(std::pair<int, int> const&, folly::EventBase*, bool, unsigned int)::{lambda(auto:1*)#2}::operator()<nebula::raftex::cpp2::RaftexServiceAsyncClient>(nebula::raftex::cpp2::RaftexServiceAsyncClient*) const src/common/thrift/ThriftClientManager.inl:62
    #1 0xb30eaf in std::_Sp_counted_deleter<nebula::raftex::cpp2::RaftexServiceAsyncClient*, nebula::thrift::ThriftClientManager<nebula::raftex::cpp2::RaftexServiceAsyncClient>::client(std::pair<int, int> const&, folly::EventBase*, bool, unsigned int)::{lambda(auto:1*)#2}, std::allocator<void>, (__gnu_cxx::_Lock_policy)2>::_M_dispose() /usr/include/c++/8/bits/shared_ptr_base.h:471
    #2 0x9200e8 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() /usr/include/c++/8/bits/shared_ptr_base.h:155
    #3 0x91da33 in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count() /usr/include/c++/8/bits/shared_ptr_base.h:728
    #4 0xafec69 in std::__shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr() /usr/include/c++/8/bits/shared_ptr_base.h:1167
    #5 0xafecca in std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient>::~shared_ptr() /usr/include/c++/8/bits/shared_ptr.h:103
    #6 0xb2aa01 in std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> >::~pair() /usr/include/c++/8/bits/stl_pair.h:208
    #7 0xb2aa50 in void __gnu_cxx::new_allocator<std::__detail::_Hash_node<std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> >, true> >::destroy<std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> > >(std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> >*) /usr/include/c++/8/ext/new_allocator.h:140
    #8 0xb26c98 in void std::allocator_traits<std::allocator<std::__detail::_Hash_node<std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> >, true> > >::destroy<std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> > >(std::allocator<std::__detail::_Hash_node<std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> >, true> >&, std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> >*) /usr/include/c++/8/bits/alloc_traits.h:487
    #9 0xb1fb99 in std::__detail::_Hashtable_alloc<std::allocator<std::__detail::_Hash_node<std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> >, true> > >::_M_deallocate_node(std::__detail::_Hash_node<std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> >, true>*) /usr/include/c++/8/bits/hashtable_policy.h:2100
    #10 0xb1d908 in std::__detail::_Hashtable_alloc<std::allocator<std::__detail::_Hash_node<std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> >, true> > >::_M_deallocate_nodes(std::__detail::_Hash_node<std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> >, true>*) /usr/include/c++/8/bits/hashtable_policy.h:2113
    #11 0xb17a06 in std::_Hashtable<std::pair<std::pair<int, int>, folly::EventBase*>, std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> >, std::allocator<std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> > >, std::__detail::_Select1st, std::equal_to<std::pair<std::pair<int, int>, folly::EventBase*> >, std::hash<std::pair<std::pair<int, int>, folly::EventBase*> >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::clear() /usr/include/c++/8/bits/hashtable.h:2050
    #12 0xb1092b in std::_Hashtable<std::pair<std::pair<int, int>, folly::EventBase*>, std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> >, std::allocator<std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> > >, std::__detail::_Select1st, std::equal_to<std::pair<std::pair<int, int>, folly::EventBase*> >, std::hash<std::pair<std::pair<int, int>, folly::EventBase*> >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::~_Hashtable() /usr/include/c++/8/bits/hashtable.h:1374
    #13 0xb1dd3a in std::unordered_map<std::pair<std::pair<int, int>, folly::EventBase*>, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient>, std::hash<std::pair<std::pair<int, int>, folly::EventBase*> >, std::equal_to<std::pair<std::pair<int, int>, folly::EventBase*> >, std::allocator<std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> > > >::~unordered_map() /usr/include/c++/8/bits/unordered_map.h:102
    #14 0xb250fc in folly::threadlocal_detail::ElementWrapper::set<std::unordered_map<std::pair<std::pair<int, int>, folly::EventBase*>, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient>, std::hash<std::pair<std::pair<int, int>, folly::EventBase*> >, std::equal_to<std::pair<std::pair<int, int>, folly::EventBase*> >, std::allocator<std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> > > >*>(std::unordered_map<std::pair<std::pair<int, int>, folly::EventBase*>, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient>, std::hash<std::pair<std::pair<int, int>, folly::EventBase*> >, std::equal_to<std::pair<std::pair<int, int>, folly::EventBase*> >, std::allocator<std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> > > >*)::{lambda(void*, folly::TLPDestructionMode)#2}::operator()(void*, folly::TLPDestructionMode) const /opt/nebula/third-party/folly/include/folly/detail/ThreadLocalDetail.h:151
    #15 0xb25134 in folly::threadlocal_detail::ElementWrapper::set<std::unordered_map<std::pair<std::pair<int, int>, folly::EventBase*>, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient>, std::hash<std::pair<std::pair<int, int>, folly::EventBase*> >, std::equal_to<std::pair<std::pair<int, int>, folly::EventBase*> >, std::allocator<std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> > > >*>(std::unordered_map<std::pair<std::pair<int, int>, folly::EventBase*>, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient>, std::hash<std::pair<std::pair<int, int>, folly::EventBase*> >, std::equal_to<std::pair<std::pair<int, int>, folly::EventBase*> >, std::allocator<std::pair<std::pair<std::pair<int, int>, folly::EventBase*> const, std::shared_ptr<nebula::raftex::cpp2::RaftexServiceAsyncClient> > > >*)::{lambda(void*, folly::TLPDestructionMode)#2}::_FUN(void*, folly::TLPDestructionMode) /opt/nebula/third-party/folly/include/folly/detail/ThreadLocalDetail.h:150
    #16 0x107761b in folly::threadlocal_detail::ElementWrapper::dispose(folly::TLPDestructionMode) /home/laura.ding/src/nebula-3rdparty/third-party/folly/_build/folly-2018.08.20.00/folly/detail/ThreadLocalDetail.h:127
    #17 0x107761b in folly::threadlocal_detail::StaticMetaBase::onThreadExit(void*) /home/laura.ding/src/nebula-3rdparty/third-party/folly/_build/folly-2018.08.20.00/folly/detail/ThreadLocalDetail.cpp:143
    #18 0x7f5678420740 in __nptl_deallocate_tsd /usr/src/debug/glibc-2.28/nptl/pthread_create.c:301
    #19 0x7f56784215ae in __nptl_deallocate_tsd /usr/src/debug/glibc-2.28/nptl/pthread_create.c:256
    #20 0x7f56784215ae in start_thread /usr/src/debug/glibc-2.28/nptl/pthread_create.c:497
    #21 0x7f567834e512 in clone (/lib64/libc.so.6+0xfd512)

0x616000330080 is located 0 bytes inside of 608-byte region [0x616000330080,0x6160003302e0)
freed by thread T145 (ThriftIO0) here:
    #0 0x7f567974ca50 in operator delete(void*, unsigned long) (/lib64/libasan.so.5+0xf2a50)
    #1 0x113708d in folly::threadlocal_detail::ElementWrapper::set<folly::EventBaseManager::EventBaseInfo*>(folly::EventBaseManager::EventBaseInfo*)::{lambda(void*, folly::TLPDestructionMode)#2}::operator()(void*, folly::TLPDestructionMode) const /home/laura.ding/src/nebula-3rdparty/third-party/folly/_build/folly-2018.08.20.00/folly/detail/ThreadLocalDetail.h:151
    #2 0x113708d in folly::threadlocal_detail::ElementWrapper::set<folly::EventBaseManager::EventBaseInfo*>(folly::EventBaseManager::EventBaseInfo*)::{lambda(void*, folly::TLPDestructionMode)#2}::_FUN(void*, folly::TLPDestructionMode) /home/laura.ding/src/nebula-3rdparty/third-party/folly/_build/folly-2018.08.20.00/folly/detail/ThreadLocalDetail.h:150
    #3 0x1135933 in folly::threadlocal_detail::ElementWrapper::dispose(folly::TLPDestructionMode) /home/laura.ding/src/nebula-3rdparty/third-party/folly/_build/folly-2018.08.20.00/folly/detail/ThreadLocalDetail.h:127
    #4 0x1135933 in folly::ThreadLocalPtr<folly::EventBaseManager::EventBaseInfo, void, void>::reset(folly::EventBaseManager::EventBaseInfo*) /home/laura.ding/src/nebula-3rdparty/third-party/folly/_build/folly-2018.08.20.00/folly/ThreadLocal.h:189
    #5 0x1135933 in folly::EventBaseManager::clearEventBase() /home/laura.ding/src/nebula-3rdparty/third-party/folly/_build/folly-2018.08.20.00/folly/io/async/EventBaseManager.cpp:60
    #6 0x1079751 in folly::IOThreadPoolExecutor::threadRun(std::shared_ptr<folly::ThreadPoolExecutor::Thread>) /home/laura.ding/src/nebula-3rdparty/third-party/folly/_build/folly-2018.08.20.00/folly/executors/IOThreadPoolExecutor.cpp:181
    #7 0x1083f4a in void std::__invoke_impl<void, void (folly::ThreadPoolExecutor::*&)(std::shared_ptr<folly::ThreadPoolExecutor::Thread>), folly::ThreadPoolExecutor*&, std::shared_ptr<folly::ThreadPoolExecutor::Thread>&>(std::__invoke_memfun_deref, void (folly::ThreadPoolExecutor::*&)(std::shared_ptr<folly::ThreadPoolExecutor::Thread>), folly::ThreadPoolExecutor*&, std::shared_ptr<folly::ThreadPoolExecutor::Thread>&) /usr/include/c++/8/bits/invoke.h:73
    #8 0x1083f4a in std::__invoke_result<void (folly::ThreadPoolExecutor::*&)(std::shared_ptr<folly::ThreadPoolExecutor::Thread>), folly::ThreadPoolExecutor*&, std::shared_ptr<folly::ThreadPoolExecutor::Thread>&>::type std::__invoke<void (folly::ThreadPoolExecutor::*&)(std::shared_ptr<folly::ThreadPoolExecutor::Thread>), folly::ThreadPoolExecutor*&, std::shared_ptr<folly::ThreadPoolExecutor::Thread>&>(void (folly::ThreadPoolExecutor::*&)(std::shared_ptr<folly::ThreadPoolExecutor::Thread>), folly::ThreadPoolExecutor*&, std::shared_ptr<folly::ThreadPoolExecutor::Thread>&) /usr/include/c++/8/bits/invoke.h:95
    #9 0x1083f4a in void std::_Bind<void (folly::ThreadPoolExecutor::*(folly::ThreadPoolExecutor*, std::shared_ptr<folly::ThreadPoolExecutor::Thread>))(std::shared_ptr<folly::ThreadPoolExecutor::Thread>)>::__call<void, , 0ul, 1ul>(std::tuple<>&&, std::_Index_tuple<0ul, 1ul>) /usr/include/c++/8/functional:400
    #10 0x1083f4a in void std::_Bind<void (folly::ThreadPoolExecutor::*(folly::ThreadPoolExecutor*, std::shared_ptr<folly::ThreadPoolExecutor::Thread>))(std::shared_ptr<folly::ThreadPoolExecutor::Thread>)>::operator()<, void>() /usr/include/c++/8/functional:484
    #11 0x1083f4a in void folly::detail::function::FunctionTraits<void ()>::callBig<std::_Bind<void (folly::ThreadPoolExecutor::*(folly::ThreadPoolExecutor*, std::shared_ptr<folly::ThreadPoolExecutor::Thread>))(std::shared_ptr<folly::ThreadPoolExecutor::Thread>)> >(folly::detail::function::Data&) /home/laura.ding/src/nebula-3rdparty/third-party/folly/_build/folly-2018.08.20.00/folly/Function.h:319
    #12 0x948b96 in folly::detail::function::FunctionTraits<void ()>::operator()() /opt/nebula/third-party/folly/include/folly/Function.h:328
    #13 0xb3fb22 in folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}::operator()() /opt/nebula/third-party/folly/include/folly/executors/thread_factory/NamedThreadFactory.h:40
    #14 0xb42f5a in void std::__invoke_impl<void, folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}>(std::__invoke_other, folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}&&) /usr/include/c++/8/bits/invoke.h:60
    #15 0xb41061 in std::__invoke_result<folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}>::type std::__invoke<folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}>(std::__invoke_result&&, (folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}&&)...) /usr/include/c++/8/bits/invoke.h:95
    #16 0xb57868 in decltype (__invoke((_S_declval<0ul>)())) std::thread::_Invoker<std::tuple<folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) /usr/include/c++/8/thread:244
    #17 0xb57612 in std::thread::_Invoker<std::tuple<folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}> >::operator()() /usr/include/c++/8/thread:253
    #18 0xb57501 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}> > >::_M_run() /usr/include/c++/8/thread:196
    #19 0x12936a2 in execute_native_thread_routine (/data/pagefault/workspace/nebula/build/src/kvstore/raftex/test/_build/leader_election_test+0x12936a2)
    #20 0x7f567842158d in start_thread /usr/src/debug/glibc-2.28/nptl/pthread_create.c:486
    #21 0x7f567834e512 in clone (/lib64/libc.so.6+0xfd512)

previously allocated by thread T145 (ThriftIO0) here:
    #0 0x7f567974b470 in operator new(unsigned long) (/lib64/libasan.so.5+0xf1470)
    #1 0x113634e in folly::EventBaseManager::EventBaseInfo::EventBaseInfo() /home/laura.ding/src/nebula-3rdparty/third-party/folly/_build/folly-2018.08.20.00/folly/io/async/EventBaseManager.h:115
    #2 0x113634e in folly::EventBaseManager::getEventBase() const /home/laura.ding/src/nebula-3rdparty/third-party/folly/_build/folly-2018.08.20.00/folly/io/async/EventBaseManager.h:115
    #3 0x107942e in folly::IOThreadPoolExecutor::threadRun(std::shared_ptr<folly::ThreadPoolExecutor::Thread>) /home/laura.ding/src/nebula-3rdparty/third-party/folly/_build/folly-2018.08.20.00/folly/executors/IOThreadPoolExecutor.cpp:156
    #4 0x1083f4a in void std::__invoke_impl<void, void (folly::ThreadPoolExecutor::*&)(std::shared_ptr<folly::ThreadPoolExecutor::Thread>), folly::ThreadPoolExecutor*&, std::shared_ptr<folly::ThreadPoolExecutor::Thread>&>(std::__invoke_memfun_deref, void (folly::ThreadPoolExecutor::*&)(std::shared_ptr<folly::ThreadPoolExecutor::Thread>), folly::ThreadPoolExecutor*&, std::shared_ptr<folly::ThreadPoolExecutor::Thread>&) /usr/include/c++/8/bits/invoke.h:73
    #5 0x1083f4a in std::__invoke_result<void (folly::ThreadPoolExecutor::*&)(std::shared_ptr<folly::ThreadPoolExecutor::Thread>), folly::ThreadPoolExecutor*&, std::shared_ptr<folly::ThreadPoolExecutor::Thread>&>::type std::__invoke<void (folly::ThreadPoolExecutor::*&)(std::shared_ptr<folly::ThreadPoolExecutor::Thread>), folly::ThreadPoolExecutor*&, std::shared_ptr<folly::ThreadPoolExecutor::Thread>&>(void (folly::ThreadPoolExecutor::*&)(std::shared_ptr<folly::ThreadPoolExecutor::Thread>), folly::ThreadPoolExecutor*&, std::shared_ptr<folly::ThreadPoolExecutor::Thread>&) /usr/include/c++/8/bits/invoke.h:95
    #6 0x1083f4a in void std::_Bind<void (folly::ThreadPoolExecutor::*(folly::ThreadPoolExecutor*, std::shared_ptr<folly::ThreadPoolExecutor::Thread>))(std::shared_ptr<folly::ThreadPoolExecutor::Thread>)>::__call<void, , 0ul, 1ul>(std::tuple<>&&, std::_Index_tuple<0ul, 1ul>) /usr/include/c++/8/functional:400
    #7 0x1083f4a in void std::_Bind<void (folly::ThreadPoolExecutor::*(folly::ThreadPoolExecutor*, std::shared_ptr<folly::ThreadPoolExecutor::Thread>))(std::shared_ptr<folly::ThreadPoolExecutor::Thread>)>::operator()<, void>() /usr/include/c++/8/functional:484
    #8 0x1083f4a in void folly::detail::function::FunctionTraits<void ()>::callBig<std::_Bind<void (folly::ThreadPoolExecutor::*(folly::ThreadPoolExecutor*, std::shared_ptr<folly::ThreadPoolExecutor::Thread>))(std::shared_ptr<folly::ThreadPoolExecutor::Thread>)> >(folly::detail::function::Data&) /home/laura.ding/src/nebula-3rdparty/third-party/folly/_build/folly-2018.08.20.00/folly/Function.h:319
    #9 0x948b96 in folly::detail::function::FunctionTraits<void ()>::operator()() /opt/nebula/third-party/folly/include/folly/Function.h:328
    #10 0xb3fb22 in folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}::operator()() /opt/nebula/third-party/folly/include/folly/executors/thread_factory/NamedThreadFactory.h:40
    #11 0xb42f5a in void std::__invoke_impl<void, folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}>(std::__invoke_other, folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}&&) /usr/include/c++/8/bits/invoke.h:60
    #12 0xb41061 in std::__invoke_result<folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}>::type std::__invoke<folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}>(std::__invoke_result&&, (folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}&&)...) /usr/include/c++/8/bits/invoke.h:95
    #13 0xb57868 in decltype (__invoke((_S_declval<0ul>)())) std::thread::_Invoker<std::tuple<folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) /usr/include/c++/8/thread:244
    #14 0xb57612 in std::thread::_Invoker<std::tuple<folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}> >::operator()() /usr/include/c++/8/thread:253
    #15 0xb57501 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}> > >::_M_run() /usr/include/c++/8/thread:196
    #16 0x12936a2 in execute_native_thread_routine (/data/pagefault/workspace/nebula/build/src/kvstore/raftex/test/_build/leader_election_test+0x12936a2)
    #17 0x7f567842158d in start_thread /usr/src/debug/glibc-2.28/nptl/pthread_create.c:486
    #18 0x7f567834e512 in clone (/lib64/libc.so.6+0xfd512)

Thread T145 (ThriftIO0) created by T96 here:
    #0 0x7f56796a6043 in __interceptor_pthread_create (/lib64/libasan.so.5+0x4c043)
    #1 0x1293778 in std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) (/data/pagefault/workspace/nebula/build/src/kvstore/raftex/test/_build/leader_election_test+0x1293778)

Thread T96 created by T0 here:
    #0 0x7f56796a6043 in __interceptor_pthread_create (/lib64/libasan.so.5+0x4c043)
    #1 0x1293778 in std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) (/data/pagefault/workspace/nebula/build/src/kvstore/raftex/test/_build/leader_election_test+0x1293778)
    #2 0xabfff2 in nebula::raftex::RaftexService::start() src/kvstore/raftex/RaftexService.cpp:41
    #3 0x932722 in nebula::raftex::setupRaft(int, nebula::fs::TempDir&, std::shared_ptr<nebula::thread::GenericThreadPool>&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, std::vector<std::pair<int, int>, std::allocator<std::pair<int, int> > >&, std::vector<std::shared_ptr<nebula::raftex::RaftexService>, std::allocator<std::shared_ptr<nebula::raftex::RaftexService> > >&, std::vector<std::shared_ptr<nebula::raftex::test::TestShard>, std::allocator<std::shared_ptr<nebula::raftex::test::TestShard> > >&, std::shared_ptr<nebula::raftex::test::TestShard>&, std::vector<bool, std::allocator<bool> >) src/kvstore/raftex/test/RaftexTestBase.cpp:176
    #4 0x9106bb in nebula::raftex::LeaderElection_ElectionWithThreeCopies_Test::TestBody() src/kvstore/raftex/test/LeaderElectionTest.cpp:31
    #5 0x101ae09 in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (/data/pagefault/workspace/nebula/build/src/kvstore/raftex/test/_build/leader_election_test+0x101ae09)

SUMMARY: AddressSanitizer: heap-use-after-free src/common/thrift/ThriftClientManager.inl:62 in auto nebula::thrift::ThriftClientManager<nebula::raftex::cpp2::RaftexServiceAsyncClient>::client(std::pair<int, int> const&, folly::EventBase*, bool, unsigned int)::{lambda(auto:1*)#2}::operator()<nebula::raftex::cpp2::RaftexServiceAsyncClient>(nebula::raftex::cpp2::RaftexServiceAsyncClient*) const
Shadow bytes around the buggy address:
  0x0c2c8005dfc0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c8005dfd0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c8005dfe0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c8005dff0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c8005e000: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c2c8005e010:[fd]fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2c8005e020: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2c8005e030: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2c8005e040: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2c8005e050: fd fd fd fd fd fd fd fd fd fd fd fd fa fa fa fa
  0x0c2c8005e060: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==36593==ABORTING
```

The reason for crash here is that evb(folly::EventBase*) will be released before client( folly::ThreadLocal<ClientMap> clientMap_;).

The relevant code is as follows:

https://github.com/vesoft-inc/nebula/blob/1e2ef26f69b887cd26a74ed3ac784b73bacc801e/src/common/thrift/ThriftClientManager.inl#L18-L66

**Comment by dangleptr** [comment]
Maybe we could release client instance manually when we stop the whole raft service.

**Comment by dangleptr** [comment]
It will be fixed in 2.0
